File: PrimsKrushakal.py
import heapq

def prim_mst(graph, start=0):
    mst_cost = 0
    visited = [False] * len(graph)
    min_heap = [(0, start)]  # (cost, vertex)

    while min_heap:
        cost, u = heapq.heappop(min_heap)
        if visited[u]:
            continue
        mst_cost += cost
        visited[u] = True
        for v, weight in graph[u]:
            if not visited[v]:
                heapq.heappush(min_heap, (weight, v))
                
    return mst_cost


class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        rootU = self.find(u)
        rootV = self.find(v)
        if rootU != rootV:
            if self.rank[rootU] > self.rank[rootV]:
                self.parent[rootV] = rootU
            elif self.rank[rootU] < self.rank[rootV]:
                self.parent[rootU] = rootV
            else:
                self.parent[rootV] = rootU
                self.rank[rootU] += 1

def kruskal_mst(n, edges):
    edges.sort(key=lambda x: x[2])  # Sort by weight
    uf = UnionFind(n)
    mst_cost = 0
    for u, v, weight in edges:
        if uf.find(u) != uf.find(v):
            uf.union(u, v)
            mst_cost += weight
    return mst_cost

File: PrimsKrushakal.java
import java.util.*;

class UnionFind {
    private int[] parent, rank;

    public UnionFind(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    public int find(int u) {
        if (parent[u] != u) parent[u] = find(parent[u]);
        return parent[u];
    }

    public void union(int u, int v) {
        int rootU = find(u), rootV = find(v);
        if (rootU != rootV) {
            if (rank[rootU] > rank[rootV]) parent[rootV] = rootU;
            else if (rank[rootU] < rank[rootV]) parent[rootU] = rootV;
            else {
                parent[rootV] = rootU;
                rank[rootU]++;
            }
        }
    }
}

class Kruskal {
    public static int kruskalMST(int n, int[][] edges) {
        Arrays.sort(edges, Comparator.comparingInt(a -> a[2])); // Sort by weight
        UnionFind uf = new UnionFind(n);
        int mstCost = 0;

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], weight = edge[2];
            if (uf.find(u) != uf.find(v)) {
                uf.union(u, v);
                mstCost += weight;
            }
        }
        return mstCost;
    }
}



import java.util.*;

class Prim {
    public static int primMST(List<List<int[]>> graph) {
        int mstCost = 0;
        boolean[] visited = new boolean[graph.size()];
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        minHeap.add(new int[]{0, 0}); // {cost, vertex}

        while (!minHeap.isEmpty()) {
            int[] edge = minHeap.poll();
            int cost = edge[0], u = edge[1];
            if (visited[u]) continue;
            mstCost += cost;
            visited[u] = true;
            for (int[] neighbor : graph.get(u)) {
                int v = neighbor[0], weight = neighbor[1];
                if (!visited[v]) {
                    minHeap.add(new int[]{weight, v});
                }
            }
        }
        return mstCost;
    }
}
